README.md
# Fixed width text parser to array

Parsing different defined text formats to array data

**Parser** - parse anything according to the format definition

Format definition:
 
```python

definition = [
    ['FIELD_1', 0, 1, 'string'],
    ['FIELD_2', 1, 10, 'float'],
    ['FIELD_3', 11, 10, 'float'],
    ['FIELD_4', 23, 1, 'integer'],
    ['FIELD_5', 24, 2, 'string'],
    ['FIELD_6', 26, 4, 'integer', 0],
    ['FIELD_7', 30, 4, 'integer', None],
    ['FIELD_8', 34, 4, 'float', None],
    ['FIELD_9', 38, 4, 'float', 0.0],
]
```

Field definition ```['FIELD_6', 26, 4, 'integer', 0]``` 
* 'FIELD_6' - field name
* 26 - index of text line where the field starts (count starts from 0)
* 4 - field length in characters 
* string - field value type (allowed types: string, integer, float)
* 0 - default value if empty in text line record

For every data type default value is always None if field results as empty string.
For every data type it is possible to set default value in format definition.

### Specific format parsers
* **Seismic.SpsParser.Sps21Parser** - SPS format version num.     SPS 2.1, JAN2006
* **Seismic.VibratorParser.ApsParser** 

### File formats

* [SPS](doc/SPS.md)
* [Vibrator](doc/Vibrator.md)


### References
* [SPS - Shell Processing Support](https://en.wikipedia.org/wiki/Shell_Processing_Support)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2021-01-31
### Added
- Sps21Parser::format_point - implemented for: type, line, point, idx, code, uph time, easting, northing, elevation, day of year & time

## [0.3.2] - 2021-01-02
### Added
- SpsParser extends Parser
### Changed  
- Sps21Parser extends SpsParser

## [0.2.0] - 2020-10-27
### Changed
- format definition changed from array to dictionary

## [0.1.1] - 2020-10-22
### Added
- Parser
- Seismic/SpsParser
- Seismic/VibratorParser



test_parser.py
import unittest
from parser import Parser

definition = [
    ['FIELD_1', 0, 1, 'string'],
    ['FIELD_2', 1, 10, 'float'],
    ['FIELD_3', 11, 10, 'float'],
    ['FIELD_4', 23, 1, 'integer'],
    ['FIELD_5', 24, 2, 'string'],
    ['FIELD_6', 26, 4, 'integer', 0],
    ['FIELD_7', 30, 4, 'integer', None],
    ['FIELD_8', 34, 4, 'float', None],
    ['FIELD_9', 38, 4, 'float', 0.0],
    ['FIELD_10', 42, 2, 'string', None],
    ['FIELD_11', 44, 2, 'string', ''],
]


class ParserTest(unittest.TestCase):
    def test_parse_empty_format(self):
        parser = Parser()
        record = 'S  21528.00  27830.00  1P1             0       756755.8 2561875.5 138.1265120558'

        with self.assertRaises(Exception):
            parser.parse(record)

    def test_parse(self):
        parser = Parser(definition)
        record = 'S  21528.00  27830.00  1P1             0       756755.8 2561875.5 138.1265120558'

        data = parser.parse(record)

        self.assertEqual('S', data[0])
        self.assertEqual(21528.00, data[1])
        self.assertEqual(27830.00, data[2])
        self.assertEqual(1, data[3])
        self.assertEqual('P1', data[4])
        self.assertEqual(0, data[5])
        self.assertEqual(None, data[6])
        self.assertEqual(None, data[7])
        self.assertEqual(0.0, data[8])
        self.assertEqual(None, data[9])
        self.assertEqual('', data[10])


    def test_csv_from_array(self):
        parser = Parser(definition)

        data = parser.get_fields()
        csv = parser.csv_from_array(data)
        self.assertEqual('FIELD_1,FIELD_2,FIELD_3,FIELD_4,FIELD_5,FIELD_6,FIELD_7,FIELD_8,FIELD_9,FIELD_10,FIELD_11', csv)

    def test_array_from_csv(self):
        parser = Parser(definition)

        csv = 'FIELD_1,FIELD_2,FIELD_3,FIELD_4,FIELD_5,FIELD_6,FIELD_7,FIELD_8'
        data = parser.array_from_csv(csv, ',')

        self.assertEqual('FIELD_1', data[0])
        self.assertEqual('FIELD_8', data[7])


if __name__ == '__main__':
    unittest.main()


test_sps_parser.py
import unittest
from sps_parser import Sps21Parser


class Sps21ParserTest(unittest.TestCase):
    def test_get_fields(self):
        parser = Sps21Parser()
        fields = parser.get_fields_point()

        self.assertEqual(15, len(fields))
        self.assertEqual('EASTING', fields[10])

    def test_parse_point(self):
        parser = Sps21Parser()
        record = 'S  21528.00  27830.00  1P1             0       756755.8 2561875.5 138.1265120558'

        data = parser.parse_point(record)
        self.assertEqual('S', data[0])
        self.assertEqual(21528.00, data[1])
        self.assertEqual(27830.00, data[2])
        self.assertEqual(1, data[3])
        self.assertEqual('P1', data[4])
        self.assertEqual(None, data[5])
        self.assertEqual(None, data[6])
        self.assertEqual(None, data[7])
        self.assertEqual(0, data[8])
        self.assertEqual(None, data[9])
        self.assertEqual(756755.8, data[10])
        self.assertEqual(2561875.5, data[11])
        self.assertEqual(138.1, data[12])
        self.assertEqual(265, data[13])
        self.assertEqual('120558', data[14])

    def test_parse_wrong_record_point(self):
        parser = Sps21Parser()
        record = 'B  21528.00  27830.00  1P1             0       756755.8 2561875.5 138.1265120558'

        data = parser.parse_point(record)
        self.assertIsNone(data)

    def test_parse_not_implemented(self):
        parser = Sps21Parser()
        record = 'X  21528.00  27830.00  1P1             0       756755.8 2561875.5 138.1265120558'

        with self.assertRaises(Exception):
            parser.parse(record)

    def test_format_not_implemented(self):
        parser = Sps21Parser()

        with self.assertRaises(Exception):
            parser.format([])

    def test_csv_point(self):
        parser = Sps21Parser()

        data = parser.get_fields_point()
        csv = parser.csv_from_array(data)
        self.assertEqual(
            'RECORD_ID,LINE,POINT,POINT_IDX,POINT_CODE,STATIC_COR,POINT_DEPTH,DATUM,UPHOLE_TIME,WATER_DEPTH,EASTING,NORTHING,ELEVATION,DAY_OF_YEAR,TIME',
            csv
        )

    def test_parse_relation(self):
        parser = Sps21Parser()
        record = 'X  1001   8287311  19248.00  27516.001    1  4351  27023.00  18875.00  19743.001'

        data = parser.parse_relation(record)

        self.assertEqual('X', data[0])
        self.assertEqual('1001', data[1])
        self.assertEqual(82873, data[2])
        self.assertEqual(1, data[3])
        self.assertEqual('1', data[4])
        self.assertEqual(19248.00, data[5])
        self.assertEqual(27516.00, data[6])
        self.assertEqual(1, data[7])
        self.assertEqual(1, data[8])
        self.assertEqual(435, data[9])
        self.assertEqual(1, data[10])
        self.assertEqual(27023.00, data[11])
        self.assertEqual(18875.00, data[12])
        self.assertEqual(19743.00, data[13])
        self.assertEqual(1, data[14])

    def test_parse_wrong_record_relation(self):
        parser = Sps21Parser()
        record = 'B  1001   8287311  19248.00  27516.001    1  4351  27023.00  18875.00  19743.001'

        data = parser.parse_relation(record)
        self.assertIsNone(data)

    def test_csv_relation(self):
        parser = Sps21Parser()

        data = parser.get_fields_relation()
        csv = parser.csv_from_array(data)

        self.assertEqual(
            'RECORD_ID,TAPE_NUMBER,RECORD_NUMBER,RECORD_INCREMENT,INSTRUMENT_CODE,S_LINE,POINT,POINT_IDX,FROM_CHANNEL,TO_CHANNEL,CHANNEL_INCREMENT,R_LINE,FROM_RECEIVER,TO_RECEIVER,RECEIVER_IDX',
            csv
        )


if __name__ == '__main__':
    unittest.main()


tests/test_parser.py
import unittest
from Parser.Parser import Parser

definition = [
    ['FIELD_1', 0, 1, 'string'],
    ['FIELD_2', 1, 10, 'float'],
    ['FIELD_3', 11, 10, 'float'],
    ['FIELD_4', 23, 1, 'integer'],
    ['FIELD_5', 24, 2, 'string'],
    ['FIELD_6', 26, 4, 'integer', 0],
    ['FIELD_7', 30, 4, 'integer', None],
    ['FIELD_8', 34, 4, 'float', None],
    ['FIELD_9', 38, 4, 'float', 0.0],
    ['FIELD_10', 42, 2, 'string', None],
    ['FIELD_11', 44, 2, 'string', ''],
]


class ParserTest(unittest.TestCase):
    def test_parse_empty_format(self):
        parser = Parser()
        record = 'S  21528.00  27830.00  1P1             0       756755.8 2561875.5 138.1265120558'

        with self.assertRaises(Exception):
            parser.parse(record)

    def test_parse(self):
        parser = Parser(definition)
        record = 'S  21528.00  27830.00  1P1             0       756755.8 2561875.5 138.1265120558'

        data = parser.parse(record)

        self.assertEqual('S', data[0])
        self.assertEqual(21528.00, data[1])
        self.assertEqual(27830.00, data[2])
        self.assertEqual(1, data[3])
        self.assertEqual('P1', data[4])
        self.assertEqual(0, data[5])
        self.assertEqual(None, data[6])
        self.assertEqual(None, data[7])
        self.assertEqual(0.0, data[8])
        self.assertEqual(None, data[9])
        self.assertEqual('', data[10])


    def test_csv_from_array(self):
        parser = Parser(definition)

        data = parser.get_fields()
        csv = parser.csv_from_array(data)
        self.assertEqual('FIELD_1,FIELD_2,FIELD_3,FIELD_4,FIELD_5,FIELD_6,FIELD_7,FIELD_8,FIELD_9,FIELD_10,FIELD_11', csv)

    def test_array_from_csv(self):
        parser = Parser(definition)

        csv = 'FIELD_1,FIELD_2,FIELD_3,FIELD_4,FIELD_5,FIELD_6,FIELD_7,FIELD_8'
        data = parser.array_from_csv(csv, ',')

        self.assertEqual('FIELD_1', data[0])
        self.assertEqual('FIELD_8', data[7])


if __name__ == '__main__':
    unittest.main()


parser.py
"""
    Parser
"""


class Parser:
    """
        Parser
    """

    def __init__(self, definition=None):
        """
        Text line format definition
        """
        self.definition = definition

    def set_definition(self, definition):
        """
        Set definition
        :param definition:
        :return:
        """
        self.definition = definition

    def parse(self, text_line):
        """
        Parse text line record
        :param text_line:
        :return:
        """
        self.is_definition()
        data = []
        for field in self.definition:
            parsed_val = self.substr(text_line, field[1], field[2]).strip()
            if len(field) == 5:
                data.append(self.convert(field[3], parsed_val, field[4]))
            else:
                data.append(self.convert(field[3], parsed_val))

        return data

    def get_fields(self):
        """
        Get definition fields name
        :return:
        """
        self.is_definition()
        fields = []
        for field in self.definition:
            fields.append(field[0])

        return fields

    def csv_from_array(self, array_data, delimiter=','):
        """
        Create CSV from array data
        :param array_data:
        :param delimiter:
        :return:
        """
        return delimiter.join(array_data)

    def array_from_csv(self, csv, delimiter=','):
        """
        Create array data from CSV
        :param csv:
        :param delimiter:
        :return:
        """
        return csv.split(delimiter)

    def parse_int(self, value, default=None):
        """
        Parse integer from string
        :param value:
        :param default:
        :return:
        """
        try:
            return int(float(value.strip()))
        except:
            return default

    def parse_float(self, value, default=None):
        """
        Parse float from string
        :param value:
        :param default:
        :return:
        """
        try:
            return float(value.strip())
        except:
            return default

    def substr(self, text, start, length):
        """
        :param text:
        :param start:
        :param length:
        :return:
        """
        return text[start:start + length]

    def convert(self, value_type, string_value, default_value=None):
        """
        Converts string value to value as per value_type
        :param value_type:
        :param string_value:
        :param default_value:
        :return:
        """
        if value_type == 'string':
            return string_value
        elif value_type == 'integer':
            return self.parse_int(string_value, default_value)
        elif value_type == 'float':
            return self.parse_float(string_value, default_value)
        else:
            return string_value

    def is_definition(self):
        """
        Check if format definition is set
        :return:
        """
        if self.definition is None:
            raise Exception('Text line format definition missing !')

    # *** TO OVERRIDE ***
    def format(self, array_data):
        """
        Converts array to text line according to the definition
        """
        raise NotImplementedError()


FixedWidthTextParser/Seismic/SpsParser.py
"""
    SPS Parser
"""
from FixedWidthTextParser.Parser import Parser


class Point:
    def __init__(self, data_array):
        self.type = data_array[0]
        self.line = data_array[1]
        self.point = data_array[2]
        self.point_idx = data_array[3]
        self.point_code = data_array[4]
        self.static_cor = data_array[5]
        self.point_depth = data_array[6]
        self.datum = data_array[7]
        self.uphole_time = data_array[8]
        self.water_depth = data_array[9]
        self.easting = data_array[10]
        self.northing = data_array[11]
        self.elevation = data_array[12]
        self.day_of_year = data_array[13]
        self.time = data_array[14]


class Relation:
    def __init__(self, data_array):
        self.type = data_array[0]
        self.tape = data_array[1]
        self.ffid = data_array[2]
        self.ffid_increment = data_array[3]
        self.instrument = data_array[4]
        self.line = data_array[5]
        self.point = data_array[6]
        self.point_idx = data_array[7]
        self.from_channel = data_array[8]
        self.to_channel = data_array[9]
        self.channel_increment = data_array[10]
        self.rcv_line = data_array[11]
        self.from_rcv = data_array[12]
        self.to_rcv = data_array[13]
        self.rcv_idx = data_array[14]


class SpsParser(Parser):
    def __init__(self, point_def: dict, relation_def: dict):
        super().__init__()
        self._point_def = point_def
        self._relation_def = relation_def

    def parse_point(self, text_line):
        """
        Parser SPS Point record
        :param text_line:
        :return:
        """
        record_type = self.substr(text_line, self._point_def['RECORD_ID'][0], self._point_def['RECORD_ID'][1]).strip()
        if record_type not in (SRC_DATA_RECORD, RCV_DATA_RECORD):
            return
        self.set_definition(self._point_def)
        return self.parse(text_line)

    def parse_point2obj(self, text_line):
        data = self.parse_point(text_line)

        if data is not None:
            return Point(data)
        else:
            return

    def parse_relation(self, text_line):
        """
        Parser SPS Relation record
        :param text_line:
        :return:
        """
        record_type = self.substr(text_line, self._point_def['RECORD_ID'][0], self._point_def['RECORD_ID'][1]).strip()
        if REL_DATA_RECORD != record_type:
            return

        self.set_definition(self._relation_def)
        return self.parse(text_line)

    def get_fields_point(self):
        """
        Get Point fields description
        :return:
        """
        self.set_definition(self._point_def)
        return self.get_fields()

    def get_fields_relation(self):
        """
        Get Relation fields description
        :return:
        """
        self.set_definition(self._relation_def)
        return self.get_fields()


SRC_DATA_RECORD = 'S'
RCV_DATA_RECORD = 'R'
REL_DATA_RECORD = 'X'
HEADER_RECORD = 'H'


class Sps21Parser(SpsParser):
    """
        SPS version 2.1 Parser
    """
    POINT_DEF = {
        'RECORD_ID': [0, 1, 'string'],
        'LINE': [1, 10, 'float'],
        'POINT': [11, 10, 'float'],
        'POINT_IDX': [23, 1, 'integer', 1],
        'POINT_CODE': [24, 2, 'string'],
        'STATIC_COR': [26, 4, 'integer'],
        'POINT_DEPTH': [30, 4, 'float'],
        'DATUM': [34, 4, 'integer'],
        'UPHOLE_TIME': [38, 2, 'integer'],
        'WATER_DEPTH': [40, 6, 'float'],
        'EASTING': [46, 9, 'float'],
        'NORTHING': [55, 10, 'float'],
        'ELEVATION': [65, 6, 'float'],
        'DAY_OF_YEAR': [71, 3, 'integer'],
        'TIME': [74, 6, 'string'],
    }
    RELATION_DEF = {
        'RECORD_ID': [0, 1, 'string'],
        'TAPE_NUMBER': [1, 6, 'string'],
        'RECORD_NUMBER': [7, 8, 'integer'],
        'RECORD_INCREMENT': [15, 1, 'integer', 1],
        'INSTRUMENT_CODE': [16, 1, 'string', '1'],
        'S_LINE': [17, 10, 'float'],
        'POINT': [27, 10, 'float'],
        'POINT_IDX': [37, 1, 'integer'],
        'FROM_CHANNEL': [38, 5, 'integer'],
        'TO_CHANNEL': [43, 5, 'integer'],
        'CHANNEL_INCREMENT': [48, 1, 'float'],
        'R_LINE': [49, 10, 'float'],
        'FROM_RECEIVER': [59, 10, 'float'],
        'TO_RECEIVER': [69, 10, 'float'],
        'RECEIVER_IDX': [79, 1, 'integer'],
    }

    def __init__(self, point_def=None, relation_def=None):
        if point_def is None:
            point_def = self.POINT_DEF
        if relation_def is None:
            relation_def = self.RELATION_DEF
        super().__init__(point_def, relation_def)

    def format_point(self, point: Point):
        """ get formatted sps string from Point objects"""
        pattern = "%1s%10.2f%10.2f  %1d%2s            %2d      %9.1f%10.1f%6.1f%3d%6s"

        return pattern % (
            point.type,
            point.line,
            point.point,
            point.point_idx,
            point.point_code,
            point.uphole_time if point.uphole_time is not None else 0,
            point.easting,
            point.northing,
            point.elevation,
            point.day_of_year,
            point.time
        )


class Sps00Parser(SpsParser):
    """
        SPS Parser of first version
    """
    POINT_DEF = {
        'RECORD_ID': [0, 1, 'string'],
        'LINE': [1, 16, 'float'],
        'POINT': [17, 8, 'float'],
        'POINT_IDX': [25, 1, 'integer', 1],
        'POINT_CODE': [26, 2, 'string'],
        'STATIC_COR': [28, 4, 'integer'],
        'POINT_DEPTH': [32, 4, 'float'],
        'DATUM': [36, 4, 'integer'],
        'UPHOLE_TIME': [40, 2, 'integer'],
        'WATER_DEPTH': [42, 4, 'float'],
        'EASTING': [46, 9, 'float'],
        'NORTHING': [55, 10, 'float'],
        'ELEVATION': [65, 6, 'float'],
        'DAY_OF_YEAR': [71, 3, 'integer'],
        'TIME': [74, 6, 'string'],
    }
    RELATION_DEF = {}

    def __init__(self, point_def=None, relation_def=None):
        if point_def is None:
            point_def = self.POINT_DEF
        if relation_def is None:
            relation_def = self.RELATION_DEF
        super().__init__(point_def, relation_def)


FixedWidthTextParser/Seismic/VibratorParser.py
"""
    Vibrator Parser
"""
from FixedWidthTextParser.Parser import Parser

# H26 APS Vibrator Attributes
# H26
# H26 Item	Definition of field  		Cols  	Format	Min to Max      		Default	Units
# H26 ----	---------------------		------	------	----------------		-------	-----
# H26 1   	Record identification		1-1   	%1s   	A               		None   	-
# H26 2   	Line name            		2-17  	%16.1f	Free            		None   	-
# H26 3   	Point number         		18-25 	%8.1f 	Free            		None   	-
# H26 4   	Point index          		26-26 	%1d   	1 to 9          		1      	-
# H26 5   	Vibrator fleet number		27-27 	%1d   	Free            		None   	-
# H26 6   	Vibrator number      		28-29 	%2d   	Free            		None   	-
# H26 7   	Vibrator drive level 		30-32 	%3d   	0 to 100        		None   	%
# H26 8   	Average phase        		33-36 	%4d   	-180 to 180     		None   	deg
# H26 9   	Peak phase           		37-40 	%4d   	-180 to 180     		None   	deg
# H26 10  	Average distortion   		41-42 	%2d   	0 to 99         		None   	%
# H26 11  	Peak distortion      		43-44 	%2d   	0 to 99         		None   	%
# H26 12  	Average force        		45-46 	%2d   	0 to 99         		None   	%
# H26 13  	Peak force           		47-49 	%3d   	Free            		None   	%
# H26 14  	Average ground stiffness		50-52 	%3d   	Free            		None   	-
# H26 15  	Average ground viscosity		53-55 	%3d   	Free            		None   	-
# H26 16  	Vib. position easting		56-64 	%9.1f 	Free            		None   	metre
# H26 17  	Vib. position northing		65-74 	%10.1f	Free            		None   	metre
# H26 18  	Vib. position elevation		75-80 	%6.1f 	-999.9 to 9999.9		None   	metre
# H26
# H26      1         2         3         4         5         6         7         8
# H26 5678901234567890123456789012345678901234567890123456789012345678901234567890
# H26
#
# A         19064.0 25360.01222 70   1   2101863 71 56 72 725883.0 2531118.2 121.6

aps = [
    ['RECORD_ID', 0, 1, 'string', None],
    ['LINE', 1, 16, 'float', None],
    ['POINT', 17, 8, 'float', None],
    ['POINT_IDX', 25, 1, 'integer', 1],
    ['VIB_FLEET_NO', 26, 1, 'integer', None],
    ['VIB_NO', 27, 2, 'integer', None],
    ['VIB_DRIVE_LEVEL', 29, 3, 'integer', None],
    ['PHASE_AVG', 32, 4, 'integer', None],
    ['PHASE_PEAK', 36, 4, 'integer', None],
    ['DIST_AVG', 40, 2, 'integer', None],
    ['DIST_PEAK', 42, 2, 'integer', None],
    ['FORCE_AVG', 44, 3, 'integer', None],
    ['FORCE_PEAK', 46, 3, 'integer', None],
    ['STIFF', 49, 3, 'integer', None],
    ['VISC', 52, 3, 'integer', None],
    ['EASTING', 55, 9, 'float', None],
    ['NORTHING', 64, 10, 'float', None],
    ['ELEVATION', 74, 6, 'float', None],
]


class ApsParser(Parser):
    def __init__(self):
        super().__init__()
        self.set_definition(aps)

    def parse_point(self, text_line):
        record_type = self.substr(text_line, aps[0][1], aps[0][2]).strip()

        if record_type != 'A':
            return None

        return self.parse(text_line)

    def parse_point2obj(self, text_line):
        data = self.parse_point(text_line)

        if data is not None:
            return ApsPoint(data)
        else:
            return


class ApsPoint:
    def __init__(self, data_array):
        self.type = data_array[0]
        self.line = data_array[1]
        self.point = data_array[2]
        self.point_idx = data_array[3]
        self.vib_fleet_no = data_array[4]
        self.vib_no = data_array[5]
        self.vib_drive_level = data_array[6]
        self.phase_avg = data_array[7]
        self.phase_peak = data_array[8]
        self.dist_avg = data_array[9]
        self.dist_peak = data_array[10]
        self.force_avg = data_array[11]
        self.force_peak = data_array[12]
        self.stiff = data_array[13]
        self.visc = data_array[14]
        self.easting = data_array[15]
        self.northing = data_array[16]
        self.elevation = data_array[17]


# H26 C.O.G Attributes Record Specification
# H26
# H26 Item	Definition of field  		Cols  	Format	Min to Max      		Default	Units
# H26 ----	---------------------		------	------	----------------		-------	-----
# H26 1   	Record identification		1-1   	%1s   	A               		None   	-
# H26 2   	Line name            		2-17  	%16.1f	Free            		None   	-
# H26 3   	Point number         		18-25 	%8.1f 	Free            		None   	-
# H26 4   	Point index          		26-26 	%1d   	1 to 9          		1      	-
# H26 5   	COG state            		28-28 	%1d   	0 to 7          		None   	-
# H26 6   	COG position easting 		30-38 	%9.1f 	Free            		None   	metre
# H26 7   	COG position northing		40-49 	%10.1f	Free            		None   	metre
# H26 8   	COG position elevation		51-56 	%6.1f 	Free            		None   	metre
# H26 9   	Deviation COG - Source		60-69 	%10.1f	Free            		None   	metre
# H26
# H26       items  5 : 0= No Cog
# H26                  1= Estimated Cog
# H26                  2= Est. Radial Err.
# H26                  3= Actual Cog
# H26                  4= Radial Err.
# H26                  5= Missing Position
# H26                  6= Inaccurate Cog
# H26                  7= Natural Cog
# H26
# H26
# H26      1         2         3         4         5         6         7         8
# H26 5678901234567890123456789012345678901234567890123456789012345678901234567890
# H26
#
# C         19064.0 25360.01 3  725883.0  2531118.2  121.6          2.5

cog = [
    ['RECORD_ID', 0, 1, 'string', None],
    ['LINE', 1, 16, 'float', None],
    ['POINT', 17, 8, 'float', None],
    ['POINT_IDX', 25, 1, 'integer', 1],
    ['COG_STATE', 27, 1, 'integer', None],
    ['EASTING', 29, 9, 'float', None],
    ['NORTHING', 39, 10, 'float', None],
    ['ELEVATION', 50, 6, 'float', None],
    ['DEVIATION', 59, 10, 'float', None],
]


class CogParser(Parser):
    def __init__(self):
        super().__init__()
        self.set_definition(cog)

    def parse_point(self, text_line):
        record_type = self.substr(text_line, cog[0][1], cog[0][2]).strip()

        if record_type != 'C':
            return None

        return self.parse(text_line)

    def parse_point2obj(self, text_line):
        data = self.parse_point(text_line)

        if data is not None:
            return CogPoint(data)
        else:
            return


class CogPoint:
    def __init__(self, data_array):
        self.type = data_array[0]
        self.line = data_array[1]
        self.point = data_array[2]
        self.point_idx = data_array[3]
        self.cog_state = data_array[4]
        self.easting = data_array[5]
        self.northing = data_array[6]
        self.elevation = data_array[7]
        self.deviation = data_array[8]


sps_parser.py
"""
    SPS Parser
"""
from parser import Parser

SRC_DATA_RECORD = 'S'
RCV_DATA_RECORD = 'R'
REL_DATA_RECORD = 'X'
HEADER_RECORD = 'H'

# H00 SPS format version num.     SPS 2.1, JAN2006
# H26
# H26 Item  Definition of field       Cols   Format  Min to Max     Default  Units
# H26 ----  -------------------       ----   ------  ----------     -------  -----
# H26 0     Record identification     1-1    A1      R or S         None     -
# H26 1     Line name    (right)      2-11   F10.2                  None     -
# H26 2     Point number (right)      12-21  F10.2                  None     -
# H26 3     Point index               24-24  I1      1   9          1        -
# H26 4     Point code                25-26  A2      Free           None     -
# H26 5     Static correction         27-30  I4      -999 - 999     Blank    Msec
# H26 6     Point Depth               31-34  F4.1    0 - 99.9       None     Metre
# H26 7     Seismic datum             35-38  I4      -999 - 9999    None     Metre
# H26 8     Uphole time               39-40  I2      0 - 99         Blank    Msec
# H26 9     Water depth               41-46  F6.1    0 to 9999.9    Blank    Metre
# H26 10    Map grid easting          47-55  F9.1    Free           None     metre
# H26 11    Map grid northing         56-65  F10.1   Free           None     metre
# H26 12    Map grid elevation        66-71  F6.1    -999.9 9999.9  None     metre
# H26 13    Day of Year               72-74  I3      1 999          None     -
# H26 14    Time hhmmss               75-80  3I2     000000 235959  None     -
# H26      1         2         3         4         5         6         7         8
# H26 5678901234567890123456789012345678901234567890123456789012345678901234567890
# S   3762.00   3961.00  1A2     7.2   0    64.8 454773.4 3008241.9  -0.2177042821
# S   3762.00   3959.00  1A2     7.2   0    64.7 454762.9 3008193.0  -0.2177042841
sps21point = [
    ['RECORD_ID', 0, 1, 'string'],
    ['LINE', 1, 10, 'float'],
    ['POINT', 11, 10, 'float'],
    ['POINT_IDX', 23, 1, 'integer', 1],
    ['POINT_CODE', 24, 2, 'string'],
    ['STATIC_COR', 26, 4, 'integer'],
    ['POINT_DEPTH', 30, 4, 'float'],
    ['DATUM', 34, 4, 'integer'],
    ['UPHOLE_TIME', 38, 2, 'integer'],
    ['WATER_DEPTH', 40, 6, 'float'],
    ['EASTING', 46, 9, 'float'],
    ['NORTHING', 55, 10, 'float'],
    ['ELEVATION', 65, 6, 'float'],
    ['DAY_OF_YEAR', 71, 3, 'integer'],
    ['TIME', 74, 6, 'string'],
]

# H00 SPS format version number   SPS 2.1, JAN2006
# H26 ---------------------------------------------------------------------------
# H26 Item  Definition of field       Cols   Format  Min to Max     Default Units
# H26 ----  -------------------       ----   ------  ----------     ------- -----
# H26 0     Record identification     1-1    A1      'X'            None    -
# H26 1     Field tape number (r adj) 2-7    3A2     Free           None    -
# H26 2     Field record number       8-15   I8      0-16777216     None    -
# H26 3     Field record increment    16-16  I1      1-9            1       -
# H26 4     Instrument code           17-17  A1      1-9            1       -
# H26 5     Line name (r adj)         18-27  F10.2   -999999.99 to  None    -
# H26                                                9999999.99
# H26 6     Point number (r adj)      28-37  F10.2   -999999.99 to  None    -
# H26                                                 9999999.99
# H26 7     Point index               38-38  I1      1-9            1       -
# H26 8     From channel              39-43  I5      1-99999        None    -
# H26 9     To channel                44-48  I5      1-99999        None    -
# H26 10    Channel increment         49-49  I1      1-9            None    -
# H26 11    Line name (r adj)         50-59  F10.2   -999999.99 to  None    -
# H26                                                 9999999.99
# H26 12    From receiver (r adj)     60-69  F10.2   -999999.99 to  None    -
# H26                                                 9999999.99
# H26 13    To receiver (r adj)       70-79  F10.2   no default     None    -
# H26 14    Receiver Index            80-80  I1      1-9            1       -
# X  1001   8287311  19248.00  27516.001    1  4351  27023.00  18875.00  19743.001
# X  1001   8287311  19248.00  27516.001  436  8711  27039.00  18873.00  19743.001
sps21relation = [
    ['RECORD_ID', 0, 1, 'string'],
    ['TAPE_NUMBER', 1, 6, 'string'],
    ['RECORD_NUMBER', 7, 8, 'integer'],
    ['RECORD_INCREMENT', 15, 1, 'integer', 1],
    ['INSTRUMENT_CODE', 16, 1, 'string', '1'],
    ['S_LINE', 17, 10, 'float'],
    ['POINT', 27, 10, 'float'],
    ['POINT_IDX', 37, 1, 'integer'],
    ['FROM_CHANNEL', 38, 5, 'integer'],
    ['TO_CHANNEL', 43, 5, 'integer'],
    ['CHANNEL_INCREMENT', 48, 1, 'float'],
    ['R_LINE', 49, 10, 'float'],
    ['FROM_RECEIVER', 59, 10, 'float'],
    ['TO_RECEIVER', 69, 10, 'float'],
    ['RECEIVER_IDX', 79, 1, 'integer'],
]


class Sps21Parser(Parser):
    """
        SPS version 2.1 Parser
    """

    def parse_point(self, text_line):
        """
        Parser SPS Point record
        :param text_line:
        :return:
        """
        record_type = self.substr(text_line, sps21point[0][1], sps21point[0][2]).strip()
        if record_type not in (SRC_DATA_RECORD, RCV_DATA_RECORD):
            return
        self.set_definition(sps21point)
        return self.parse(text_line)

    def parse_relation(self, text_line):
        """
        Parser SPS Relation record
        :param text_line:
        :return:
        """
        record_type = self.substr(text_line, sps21point[0][1], sps21point[0][2]).strip()
        if record_type not in REL_DATA_RECORD:
            return

        self.set_definition(sps21relation)
        return self.parse(text_line)

    def get_fields_point(self):
        """
        Get Point fields description
        :return:
        """
        self.set_definition(sps21point)
        return self.get_fields()

    def get_fields_relation(self):
        """
        Get Relation fields description
        :return:
        """
        self.set_definition(sps21relation)
        return self.get_fields()


# H00 SPS format version num.         SPS001,08OCT1990 (SHELL EP 90-2935);
# H26
# H26 Item  Definition of field       Cols   Format  Min to Max     Default  Units
# H26 ----  -------------------       ----   ------  ----------     -------  -----
# H26 1     Record identification     1-1    A1      R or S         None     -
# H26 2     Line name (left just)     2-17   4A4     Free           None     -
# H26 3     Point number              18-25  2A4     Free           None     -
# H26 4     Point index               26-26  I1      1   9          1        -
# H26 5     Point code                27-28  A2      Free           None     -
# H26 6     Static correction         29-32  I4      -999 - 999     Blank    Msec
# H26 7     Point Depth               33-36  F4.1    0 - 99.9       None     Metre
# H26 8     Seismic datum             37-40  I4      -999 - 9999    None     Metre
# H26 9     Uphole time               41-42  I2      0 - 99         Blank    Msec
# H26 10    Water depth               43-46  F4.1    0 to 99.9/999  Blank    Metre
# H26 11    Map grid easting          47-55  F9.1    Free           None     metre
# H26 12    Map grid northing         56-65  F10.1   Free           None     metre
# H26 13    Map grid elevation        66-71  F6.1    -999.9 9999.9  None     metre
# H26 14    Day of Year               72-74  I3      1 999          None     -
# H26 15    Time hhmmss               75-80  3I2     000000 235959  None     -
# H26      1         2         3         4         5         6         7         8
# H26 5678901234567890123456789012345678901234567890123456789012345678901234567890
# S3762                39611A2     7.2   0  64.8 454773.4 3008241.9  -0.2177042821
# S3762                39591A2     7.2   0  64.7 454762.9 3008193.0  -0.2177042841
sps00point = []
sps00relation = []


class Sps00Parser(Parser):
    """
        SPS Parser of first version
    """

    pass
